### ISSUE TITLE
Conflicts regarding activated versions of default gems

### ISSUE THREAD

**Issue by janko** [rubygems/rubygems#6207]
### Describe the problem as clearly as you can

In Ruby 3.2, when a default gem is required by a gem that I required in my program, conflicts arise between my installed versions of built-in versions of default gems. I have the following gems installed:

```
mail (2.8.0)
net-protocol (default: 0.2.1, 0.1.3)
timeout (default: 0.3.1, 0.3.0)
```

When I run the following ruby script:

```rb
require "mail"
```

I get an error that it cannot activate `net-protocol` 0.1.3, because version 0.2.1 is already activated. I'm confused why it tries to activate 0.1.3, since 0.2.1 is both the latest version and the default gem on Ruby 3.2.

```
<internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `require'
        from bla.rb:1:in `<main>'
<internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/mail-2.8.0/lib/mail.rb:9:in `<module:Mail>'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/mail-2.8.0/lib/mail.rb:3:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:160:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:160:in `rescue in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:149:in `require'
        from bla.rb:1:in `<main>'
<internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/net-smtp-0.3.3/lib/net/smtp.rb:20:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:96:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:96:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/mail-2.8.0/lib/mail.rb:9:in `<module:Mail>'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/mail-2.8.0/lib/mail.rb:3:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:160:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:160:in `rescue in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:149:in `require'
        from bla.rb:1:in `<main>'
/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/specification.rb:2277:in `check_version_conflict': can't activate net-protocol-0.1.3, already activated net-protocol-0.2.1 (Gem::LoadError)
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/specification.rb:1414:in `activate'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:117:in `block in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:116:in `each'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:116:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/net-smtp-0.3.3/lib/net/smtp.rb:20:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:96:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:96:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/mail-2.8.0/lib/mail.rb:9:in `<module:Mail>'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/mail-2.8.0/lib/mail.rb:3:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:160:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:160:in `rescue in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:149:in `require'
        from bla.rb:1:in `<main>'
<internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:85:in `require': cannot load such file -- mail (LoadError)
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from bla.rb:1:in `<main>'
```

### Did you try upgrading RubyGems?

Yes, I'm running on currently latest version of RubyGems.

### Post steps to reproduce the problem

You should be able to reproduce the problem by installing the mentioned versions of the gem:

```sh
$ gem install mail -v 2.8.0
$ gem install net-protocol -v 0.1.3
$ gem install timeout -v 0.3.0
```

The `mail` gem doesn't do any version locking, it just requires `net/smtp`, which triggers the `net-protocol` conflict. If I require `net/smtp` myself, `net-protocol` gets loaded just fine, so the problem arises when a gem requires a default gem.

```rb
require "net/smtp" # works just fine
require "mail"
```

Another occurrence of this problem is attempting to run `binding.irb`, which raises a conflict in loading the `timeout` default gem.

```rb
require "net/smtp"
require "mail"

binding.irb
```
```
<internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `require'
        from <internal:prelude>:4:in `irb'
        from bla.rb:4:in `<main>'
<internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/irb.rb:13:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:prelude>:4:in `irb'
        from bla.rb:4:in `<main>'
<internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:167:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/reline.rb:2:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/irb.rb:13:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:prelude>:4:in `irb'
        from bla.rb:4:in `<main>'
/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/specification.rb:2277:in `check_version_conflict': can't activate timeout-0.3.0, already activated timeout-0.3.1 (Gem::LoadError)
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/specification.rb:1414:in `activate'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:117:in `block in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:116:in `each'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:116:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/reline.rb:2:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from /Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/irb.rb:13:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.0/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:148:in `require'
        from <internal:prelude>:4:in `irb'
        from bla.rb:4:in `<main>'
```

This one cannot be fixed by requiring `timeout` manually, the following script still fails with the same error:

```rb
require "net/smtp"
require "timeout"
require "mail"

binding.irb #~> still fails with the same error
```

### Which command did you run?

I ran `ruby program.rb` with the above code snippets.

### What were you expecting to happen?

I expected the program to resolve default gems without issues.

### What actually happened?

There was a conflict in resolving the default gem versions.

### Run `gem env` and paste the output below

```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.4.1
  - RUBY VERSION: 3.2.0 (2022-12-25 patchlevel 0) [arm64-darwin22]
  - INSTALLATION DIRECTORY: /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0
  - USER INSTALLATION DIRECTORY: /Users/janko/.gem/ruby/3.2.0
  - RUBY EXECUTABLE: /Users/janko/.rbenv/versions/3.2.0/bin/ruby
  - GIT EXECUTABLE: /opt/homebrew/bin/git
  - EXECUTABLE DIRECTORY: /Users/janko/.rbenv/versions/3.2.0/bin
  - SPEC CACHE DIRECTORY: /Users/janko/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Users/janko/.rbenv/versions/3.2.0/etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - arm64-darwin-22
  - GEM PATHS:
     - /Users/janko/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0
     - /Users/janko/.gem/ruby/3.2.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => true
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /Users/janko/.rbenv/versions/3.2.0/bin
     - /opt/homebrew/Cellar/rbenv/1.2.0/libexec
     - /opt/homebrew/opt/openjdk/bin
     - /Users/janko/bin
     - /Users/janko/.rbenv/shims
     - /opt/homebrew/bin
     - /opt/homebrew/bin
     - /opt/homebrew/sbin
     - /usr/local/bin
     - /System/Cryptexes/App/usr/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - /opt/homebrew/opt/openjdk/bin
     - /Users/janko/bin
     - /Users/janko/.rbenv/shims
     - /opt/homebrew/bin
     - /opt/homebrew/sbin
     - /Users/janko/.cargo/bin
```

**Comment by deivid-rodriguez** [comment]
Indeed weird. I cannot reproduce, and I have almost the same env as you (except for using asdf instead of rbenv). Maybe other gems installed are interfering? I'm still not sure how this could happen because of what you explained above: 

> I'm confused why it tries to activate 0.1.3, since 0.2.1 is both the latest version and the default gem on Ruby 3.2.

**Comment by janko** [comment]
OK, I cannot reproduce it on a fresh Ruby installation either, so something else must have caused this. I will close the issue for now, and open it when I pinpoint the root cause.

**Comment by deivid-rodriguez** [comment]
Alright @janko, please let us know what you find!

**Comment by indirect** [comment]
My first guess would be the `rubygems-bundler` gem, which has sometimes been installed by RVM. My next guess would be the env var `RUBYGEMS_GEMDEPS` being set. Both of those things eagerly activate Bundler as part of starting any Ruby process, and could potentially cause different versions of things to be required. Beyond that, you might have a `RUBYOPT` set that requires something while booting Ruby.

If a clean install fixes it, though, sounds like you are all set for now. 👍🏻

**Comment by janko** [comment]
I was able to reproduce the issue on a fresh installation of Ruby 3.2.1:

```sh
$ gem install mail
$ ruby -r mail -e '' # doesn't error

$ gem install net-imap -v 0.3.2 # bundled gem version: 0.3.4
$ gem install date -v 3.3.1 # default gem version: 3.3.3
$ ruby -r mail -e '' # errors
```
```
[44493, #<Thread:0x00000001024fb110 run>, #<Gem::LoadError: can't activate date-3.3.1, already activated date-3.3.3>, ["/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/specification.rb:2277:in `check_version_conflict'", "/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/specification.rb:1414:in `activate'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:120:in `block in require'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:119:in `each'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:119:in `require'", "/Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:5:in `<module:Mail>'", "/Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:3:in `<top (required)>'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `require'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `rescue in require'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:152:in `require'"]]
[44493, #<Thread:0x00000001024fb110 run>, #<RuntimeError: CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true>, ["<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:170:in `ensure in require'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:170:in `require'", "/Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:5:in `<module:Mail>'", "/Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:3:in `<top (required)>'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `require'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `rescue in require'", "<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:152:in `require'"]]
<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:170:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:170:in `require'
<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:170:in `ensure in require': CRITICAL: RUBYGEMS_ACTIVATION_MONITOR.owned?: before false -> after true (RuntimeError)
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:170:in `require'
        from /Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:5:in `<module:Mail>'
        from /Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:3:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `rescue in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:152:in `require'
/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/specification.rb:2277:in `check_version_conflict': can't activate date-3.3.1, already activated date-3.3.3 (Gem::LoadError)
        from /Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/specification.rb:1414:in `activate'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:120:in `block in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:119:in `each'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:119:in `require'
        from /Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:5:in `<module:Mail>'
        from /Users/janko/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/mail-2.8.1/lib/mail.rb:3:in `<top (required)>'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `require'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:162:in `rescue in require'
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:152:in `require'
<internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:88:in `require': cannot load such file -- mail (LoadError)
        from <internal:/Users/janko/.rbenv/versions/3.2.1/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:88:in `require'
```

I believe I've only seen this issue when not using Bundler. When I looked into these gems, I didn't see any of them trying to activate a specific version of their dependencies. And it's difficult to parse from the error message what exactly is the problem; I had to install older versions of both `net-imap` and `date` in order to reproduce it.

**Comment by hsbt** [comment]
I also reproduce this.

**Comment by aeroastro** [comment]
I have found out the reason why this occurs for https://github.com/rubygems/rubygems/issues/6207#issuecomment-1425606911

There seems to be several ways to fix this issue, and I want to know what is intended and what is not intended in this scenario. Below is the detail explanation of the scenario.


### Activated `mail` and Partially Activated its Dependencies

* Since "mail" is not a standard libraries, `require "mail"` is handled (activated and required) at the rescue block of `Kernel#.require` core_ext.
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/core_ext/kernel_require.rb#L151-L157
* During activating `spec = Gem::Specification.find_by_path "mail"`, it activates "mail" and its runtime dependencies; `mini_mime`, `net-smtp`, `net-imap`, `net-pop` at `Gem.try_activate spec`
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/specification.rb#L1428-L1433
* Since there are two `net-imap` gems (bundled 0.3.4 and manually-installed 0.3.2) and there is no `gem 'net-imap', '0.3.2'` statement, `spec.runtime_dependencies.find { |dep| dep.name == 'net-imap' }.to_specs` returns two specs.
  *  https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/dependency.rb#L304-L305
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/dependency.rb#L273-L277
* When more than one specs are returned for `runtime_dependencies`, `spec.activate_dependencies` gives up recursively activating dependencies and add the dependency to the `Gem::Specification.unresolved_deps`
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/specification.rb#L1457-L1468

### Default Gem `date` (3.3.3) is Activated during `require`

* After "mail" is activated, "mail" is required by `gem_original_require` during the same rescue block.
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/core_ext/kernel_require.rb#L151-L157
* When requiring "mail", "date" is also required in the Ruby code.
  * https://github.com/mikel/mail/blob/10a4443b9d4ffa71b9ad643ad86cc23ccc99f0f3/lib/mail.rb#L5
* Since "date" is a default gem, the latest version (3.3.3) spec is activated by `gem 'date', Gem::Requirement.default_prerelease`
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/core_ext/kernel_require.rb#L44
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/core_ext/kernel_require.rb#L72-L77

### Existence of Unresolved `net-imap` Keeps `require` Running

* The existence of `net-imap` in `Gem::Specification.unresolved_deps` prevents calling normal require for `date`
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/core_ext/kernel_require.rb#L80-L86
* Though the `date` itself is already activated, it continues to find the spec from its dependents by `Gem::Specification.find_in_unresolved_tree "date"`
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/core_ext/kernel_require.rb#L114-L116

### Unresolved Tree is Searched by `contains_requirable_file?` Not by `name`

* It traverses `Gem::Specification.send(:unresolved_specs)` and returns the first trail of specs whose last spec is `spec.contains_requirable_file?('date')`
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/specification.rb#L1071-L1080
  * https://github.com/rubygems/rubygems/blob/2e94a6ae5f92a89a83bce942911f8e0525fe177b/lib/rubygems/basic_specification.rb#L333-L337
* There are two `date` specification (default 3.3.3 and manually-installed 3.3.1) and they are evaluated in the order.
* Since `date` 3.3.3 is default gem, the `date` file cannot be found in gem's path, instead;
  * `/Users/*/.rbenv/versions/3.2.2/lib/ruby/3.2.0/date.rb`
* On the other hand, since `date` 3.3.1 is manually-installed, it can be found in gem's path;
  * `/Users/*/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/date-3.3.1/lib/date.rb` 
* `Gem::Specification.find_in_unresolved_tree "date"` finally returns incorrect `3.3.1` spec and activating it results in `Gem::LoadError: can't activate date-3.3.1, already activated date-3.3.3`

In my Mac, ruby 3.2.2 is installed via `rbenv` and `rbenv/ruby-build`, then I manually enters `gem install` as in the [comment](https://github.com/rubygems/rubygems/issues/6207#issuecomment-1425606911).

**Comment by deivid-rodriguez** [comment]
I figured this out thanks to the great investigation by @aeroastro.

The problem is during `date` require, [`date_core`](https://github.com/ruby/date/blob/4007df808231f1f6eff289962a47451be9c0e6ca/lib/date.rb#L4) is not detected as part of the `date` gem, and thus short-circuited [here](https://github.com/rubygems/rubygems/blob/53604cb26e23ad9b6b81b8c7eb909f5ac5d4e5fe/lib/rubygems/core_ext/kernel_require.rb#L45-L71).

This is because RubyGems relies on the list of files provided by the default gem gemspec for figuring out if a requirable feature belongs to the gem, but requirable features provided by extensions are not currently included in this list.

Because of this, a workaround to this issue is to manually edit the default date gemspec in 3.2.1 to include `date_core`. So that instead of:

```ruby
  s.files = ["ext/date/extconf.rb".freeze, "lib/date.rb".freeze]
```

it reads like this

```ruby
  s.files = ["ext/date/extconf.rb".freeze, "date.rb".freeze, "date_core.bundle".freeze]
```

I'm working on a proper fix at https://github.com/ruby/ruby/pull/9673.

This is also the root cause of https://github.com/rubygems/rubygems/issues/7160.

**Comment by deivid-rodriguez** [comment]
This should be fixed in Ruby 3.4.0-dev now, so I'll close this. Let us know if you find issues though!

**Comment by hsbt** [comment]
FYI: I request to backport my fixes to Ruby 3.3.1.
